File: src/main/java/com/oycm/algorithm/f/prefix_sum_hash/Solution_9.scala
package com.oycm.algorithm.f.prefix_sum_hash

object Solution_9 {

  /**
   * 3026. 最大好子数组和 1817
   * https://leetcode.cn/problems/maximum-good-subarray-sum/description/
   *
   * 好子数组：[i, j] 满足 |nums(i) - nums(j)| == k
   *
   * @param nums
   * @param k
   * @return 求 nums 好子数组的最大和
   */
  def maximumSubarraySum(nums: Array[Int], k: Int): Long = {
    /*
    枚举 j 查找 nums(j) - k 和 nums(j) + k 的 前缀和 sums(j+1) - sums(i)
     */
    var ans = Long.MinValue
    var sum = 0L
    // 相同 的 nums(i) 取较小前缀和
    val map = scala.collection.mutable.Map[Int, Long]()

    for (i <- nums.indices) {
      sum += nums(i)
      val k1 = nums(i) - k
      val k2 = nums(i) + k

      if (map.contains(k1)) {
        ans = Math.max(ans, sum - map(k1) + k1)
      }
      if (map.contains(k2)) {
        ans = Math.max(ans, sum - map(k2) + k2)
      }

      if (map.contains(nums(i))) {
        map(nums(i)) = Math.min(sum, map(nums(i)))
      } else {
        map(nums(i)) = sum
      }

    }
    if (ans == Long.MinValue) {
      ans = 0
    }

    ans
  }

  def main(args: Array[String]): Unit = {
    println(maximumSubarraySum(Array(1, 2, 3, 4, 5, 6), 1))
    println(maximumSubarraySum(Array(-1, 3, 2, 4, 5), 3))
    println(maximumSubarraySum(Array(-1, -2, -3, -4), 2))
  }

}

File: src/main/java/com/oycm/month2026/june/Solution_9.java
package com.oycm.month2026.june;

public class Solution_9 {

    /**
     * 3689. <a href="https://leetcode.cn/problems/maximum-total-subarray-value-i/description/">最大子数组总值 I</a> 1371
     *
     * @param nums
     * @param k
     * @return
     */
    public long maxTotalValue(int[] nums, int k) {
        /*
        给定一个长度为 n 的整数数组 nums 和一个整数 k。
        你必须从 nums 中选择 恰好 k 个非空子数组 nums[l..r]。子数组可以重叠，同一个子数组（相同的 l 和 r）可以 被选择超过一次。
        子数组 nums[l..r] 的 值 定义为：max(nums[l..r]) - min(nums[l..r])。
        总值 是所有被选子数组的 值 之和。
        返回你能实现的 最大 可能总值。
        子数组 是数组中连续的 非空 元素序列。
         */
        /*
        1 <= n == nums.length <= 5 * 1e4
        0 <= nums[i] <= 1e9
        1 <= k <= 105
         */
        /*
        (最大值 - 最小值) * k
         */
        /*
        思考题：计算数组的最小值和最大值，最多需要多少次比大小？至多要 3n/2。
         */
        int mx = Integer.MIN_VALUE;
        int mn = Integer.MAX_VALUE;
        for (int x : nums) {
            mx = Math.max(mx, x);
            mn = Math.min(mn, x);
        }
        return (long) (mx - mn) * k;
    }

}
